### PR TITLE
TG Powernet Refactor

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Most the changes here are completely invisible to players. This probably wasn't really needed to be honest, but I originally started implementing this, when trying to "fix" the emitter problem. Turns out the emitters weren't actually a problem. Supposed that's the advantage of doing this; I have a slightly better understanding of powernets now.

Changes that you'll notice as a player:
- Powersinks will drain power a little faster, but can explode sooner
- Multitools will display current power in the network in addition to total consumption and total load
- Fixes emitters sparks appearing where they shoudln't

:cl: Fox McCloud
tweak: Powersinks drain rate increased, but they also explode a bit sooner
add: Multitools display total power, load, and free power
fix: Fix emitter sparks sparking at the location of emitters where biult as opposed to emitter current location
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/machinery.dm [hunk 1]
```diff
@@ -64,23 +64,6 @@ SUBSYSTEM_DEF(machines)
 		if(MC_TICK_CHECK)
 			return
 
-/datum/controller/subsystem/machines/proc/process_premachines(resumed = 0)
```
Comment (Fox-McCloud) [inline code/controllers/subsystem/machinery.dm]: need to kill off deferred powernet rebuilds too.

File: code/game/objects/items/devices/powersink.dm [hunk 2]
```diff
@@ -13,127 +13,137 @@
...
-	var/datum/powernet/PN			// Our powernet
+	var/drain_rate = 2000000	// amount of power to drain per tick
+	var/power_drained = 0 		// has drained this much power
+	var/max_power = 6e8		// maximum power that can be drained before exploding
+	var/mode = 0		// 0 = off, 1=clamped (off), 2=operating
+	var/admins_warned = FALSE // stop spam, only warn the admins once that we are about to boom
+
+	var/const/DISCONNECTED = 0
+	var/const/CLAMPED_OFF = 1
+	var/const/OPERATING = 2
```
Comment (Fox-McCloud) [inline code/game/objects/items/devices/powersink.dm]: TG, why.

File: code/modules/power/apc.dm [hunk 3]
```diff
@@ -1200,14 +1197,12 @@
 				autoflag = 0
 
 		// now trickle-charge the cell
-
-		if(src.attempt_charging())
```
Comment (Fox-McCloud) [inline code/modules/power/apc.dm]: look into this proc and if it's used/necessary anymore.

File: code/modules/power/cable.dm [hunk 4]
```diff
@@ -452,18 +473,19 @@ obj/structure/cable/proc/cable_color(var/colorC)
...
 	var/obj/O = P_list[1]
 	// remove the cut cable from its turf and powernet, so that it doesn't get count in propagate_network worklist
-	loc = null
+	if(remove)
+		loc = null
 	powernet.remove_cable(src) //remove the cut cable from its powernet
 
-	// queue it to rebuild
-	SSmachines.deferred_powernet_rebuilds += O
+	addtimer(CALLBACK(O, .proc/auto_propogate_cut_cable, O), 0) //so we don't rebuild the network X times when singulo/explosion destroys a line of X cables
```
Comment (Fox-McCloud) [inline code/modules/power/cable.dm]: this is the new implementation of deferred powernet rebuilds.

**Comment by KasparoVy** [comment]
Good luck! Man all these big ticket ports.

Callback city

**Comment by Fox-McCloud** [comment]
Fixed up and tested.

Just need to see if the deferred powernets we have is better then theirs or not.

**Comment by Fox-McCloud** [comment]
After extensive testing, TG's method of deferring powernet rebuilds vs ours? Inconclusive.

Theirs seems to be more expenisve, up front, which is reflected in the auto-propogate proc and cable's `Destroy` cost being nearly double.

That said, with the deferring system we already had, it meant tons of cables failed to garbage collect.

That said, I believe these are false positives of some sort, as they showed up instantaneously in the GC log, so I'm really not sure what's going on there; if they showed up after 3 minutes, then I'd be more concerned.

Any event, going to leave our method of deferring in, as it does seem to be a bit better.

**Pull request merged**